### PR TITLE
only load required entry points which improves the performance

### PR DESCRIPTION
**Pull request by dirk-thomas**
Follow up of #424.

The changes are split over four commits to ease the review.

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/9081/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/4828/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/7432/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/9019/)

### CODE REVIEW

**Comment by mjcarroll** [comment]
Sorry, I was only approving the first commit, didn't realize that it marks the whole thing as approved, still looking.

File: ros2cli/ros2cli/command/__init__.py [hunk 1]
```diff
@@ -112,3 +125,125 @@ def add_subparsers(
...
+        self.value = ''
+
+    def __getattr__(self, name):
+        return getattr(self.value, name)
+
+    def __iter__(self):
+        return self.value.__iter__()
+
+
+def add_subparsers_on_demand(
```
Comment (mjcarroll) [inline ros2cli/ros2cli/command/__init__.py]: Since `add_subparsers` is deprecated, should there be documentation here?

Comment (dirk-thomas) [reply ros2cli/ros2cli/command/__init__.py]: Good point, added in 01616f394b849dbe1e61239da04574efeffc1106.

**Comment by ivanpauno** [comment]
Cancelled hanging windows CI, here new one:
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/9​02​1/)

**Comment by dirk-thomas** [comment]
While the Windows build has a very high test failure count (75) that seems to match the latest nightly Windows Debug build: https://ci.ros2.org/view/nightly/job/nightly_win_deb/1504/

Merging...

**Comment by ivanpauno** [comment]
I noticed that autocomplete isn't working at all in master in all the ros2 cli tools.
I reverted this together with https://github.com/ros2/ros2cli/pull/437, and it started working again.

@dirk-thomas can you take a look? If not, I can try to fix it (though I don't have much context).

**Pull request merged**